NameLayerMain/src/vg/civcraft/mc/namelayer/database/GroupManagerDao.java

```diff
@@ -820,51 +871,101 @@ public synchronized void removeSubGroup(String group, String subGroup){
 		Map<PlayerType, List<PermissionType>> perms = new HashMap<PlayerType, List<PermissionType>>();
 		try {
-			getPerms.setString(1, group);
-			ResultSet set = getPerms.executeQuery();
+			getPermission.setString(1, group);
+			ResultSet set = getPermission.executeQuery();
 			while(set.next()){
 				PlayerType type = PlayerType.getPlayerType(set.getString(1));
-				String length = set.getString(2);
-				String[] multiPerms = length.split(" ");
-				List<PermissionType> listPerm = new ArrayList<PermissionType>();
-				for (String x: multiPerms)
-					listPerm.add(PermissionType.getPermissionType(x));
-				perms.put(type, listPerm);
+				List<PermissionType> listPerm = perms.get(type);
+				if (listPerm == null) {
+					listPerm = new ArrayList<PermissionType>();
+					perms.put(type, listPerm);
+				}
+				int id = set.getInt(2);
+				PermissionType perm = PermissionType.getPermission(id);
+				if (perm != null) {
+					listPerm.add(perm);
+				}
 			}
 		} catch (SQLException e) {
 			plugin.getLogger().log(Level.WARNING, "Problem getting permissions for group " + group, e);
 		}
 		return perms;
 	}
 	
-	public synchronized void updatePermissions(String group, PlayerType pType, String perms){
+	public synchronized void updatePermissions(String group, PlayerType pType, List<PermissionType> perms){
 		NameLayerPlugin.reconnectAndReintializeStatements();
 		try {
-			updatePerm.setString(1, perms);
-			updatePerm.setString(2, group);
-			updatePerm.setString(3, pType.name());
-			updatePerm.execute();
+			removePermission.setString(1, group);
+			removePermission.setString(2, pType.name());
+			removePermission.execute();
 		} catch (SQLException e) {
-			plugin.getLogger().log(Level.WARNING, "Problem updating permissions for group " + group
+			plugin.getLogger().log(Level.WARNING, "Problem removing permissions for group " + group
 					+ " on playertype " + pType.name(), e);
 		}
+		addPermission(group,pType.name(),perms);
+	}
+	
+	public synchronized void registerPermission(PermissionType perm) {
+		NameLayerPlugin.reconnectAndReintializeStatements();
+		try {
+			registerPermission.setInt(1, perm.getId());
+			registerPermission.setString(2, perm.getName());
+			registerPermission.execute();
+		} catch (SQLException e) {
+			plugin.getLogger().log(Level.WARNING, "Problem register permission " + perm.getName(), e);
+		}
+	}
+	
+	public synchronized Map<Integer, String> getPermissionMapping() {
+		NameLayerPlugin.reconnectAndReintializeStatements();
+		Map <Integer,String> perms = new TreeMap<Integer, String>();
+		try {
+			ResultSet res = getPermissionMapping.executeQuery();
+			while (res.next()) {
+				perms.put(res.getInt(1), res.getString(2));
+			}
+		}
+		catch (SQLException e) {
+			plugin.getLogger().log(Level.WARNING, "Problem getting permissions from db", e);
+		}
+		return perms;
+	}
+	
+	public synchronized void addNewDefaultPermission(List <PlayerType> playerTypes, PermissionType perm) {
```

Comment:
Perhaps the visibility is what's throwing me off. Is it only used internally? If so, perhaps public isn't the wrote scope considering the potential impact.

What's the likelihood of someone with bindings to Namelayer calling that method?